timestamp-impl/src/main/java/com/palantir/timestamp/AtomicTimestamp.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.timestamp;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class AtomicTimestamp {
+
+    private final AtomicLong value;
+
+    public AtomicTimestamp(long initialValue) {
+        value = new AtomicLong(initialValue);
+    }
+
+    public TimestampRange incrementBy(long delta) {
+        if (delta <= 0) {
+            throw new IllegalArgumentException("Timestamp must be increased by a positive number. Found " + delta);
```

Comment:
we should probably use Preconditions to reduce bytecode size of this method as it will be on hot path. Would really like to have some JMH benchmarks to stress test some of these internals and run through [JIT Watch](https://github.com/AdoptOpenJDK/jitwatch)